exporter/elasticsearchexporter/exporter.go

```diff
@@ -65,13 +68,27 @@ func (e *elasticsearchExporter) Shutdown(ctx context.Context) error {
 	return nil
 }
 
+func (e *elasticsearchExporter) getEncoder(m MappingMode) (documentEncoder, error) {
```

Comment:
This looks vulnerable to race, while they are called in concurrent methods `push*` IIUC

Reply:
Nice find. I've switched to a sync map.
It means we need to build a new encoder every time though, to be able to use `LoadOrStore`.

Creating encoders should be cheap, so that seems fine. I could switch to a mutex though.

Reply:
Not a big deal, but I'm wondering if it would save some cycles here by being optimistic and checking whether the encoder is already in the map (`Load`) before creating it and `LoadOrStore`? Race conditions would still be possible at first execution, but I guess it would be harmless, and afterward calls would be more efficient.

Reply:
Well, if we do that, we can get rid of the `sync.Map` and use a raw map, that'd be equivalent.

Reply:
> we can get rid of the sync.Map and use a raw map, that'd be equivalent.

they'll look similar but not exactly the same, because with a raw map it'll need a mutex.

I assume `newEncoder(m)` will have 1 alloc for the later stateful pointer serializer. That might add up when the throughput is high and the stateful serializer becomes larger. This is not a blocker but we'll like it at the end of the day.

Reply:
Yes, but that mutex would be held separately in `Load` and `Store`. So a concurrent export could store data between the two, and make us override the entry afterwards.
A raw map would have exactly the same flaw.

As mentioned by @edmocosta, this would be an issue only at start. So it wouldn't be such a big issue.
We could also bootstrap the map with every known mapping type within the map.

> I assume newEncoder(m) will have 1 alloc for the later stateful pointer serializer. That might add up when the throughput is high and the stateful serializer becomes larger. This is not a blocker but we'll like it at the end of the day.

Creating one encoder for each request is the current behavior on main, and it doesn't appear to create issues.

Reply:
> We could also bootstrap the map with every known mapping type within the map.

I like it. I took some time to check whether `create*Exporter` will be called concurrently, but even so, the exporters seem to be distinct and there should be no problem.